src/app/games/lasca/lasca.component.ts

```diff
@@ -0,0 +1,330 @@
+import { LascaControlAndDominationMinimax } from './LascaControlAndDomination';
+import { LascaControlMinimax } from './LascaControlMinimax';
+import { LascaMove, LascaMoveFailure } from './LascaMove';
+import { LascaRules } from './LascaRules';
+import { LascaPiece, LascaSpace, LascaState } from './LascaState';
+import { LascaTutorial } from './LascaTutorial';
+
+class SpaceInfo {
+    spaceClasses: string[];
+    pieceInfos: LascaPieceInfo[];
+}
+class LascaPieceInfo {
+    classes: string[];
+    isOfficer: boolean;
+}
+@Component({
+    selector: 'app-lasca', // Juneau why !
+    templateUrl: './lasca.component.html',
+    styleUrls: ['../../components/game-components/game-component/game-component.scss'],
+})
+export class LascaComponent extends RectangularGameComponent<LascaRules,
+                                                             LascaMove,
+                                                             LascaState,
+                                                             LascaSpace>
+{
+    public readonly HORIZONTAL_WIDTH_RATIO: number = 1.2;
+    public readonly DIAGONALISATION: number = 0.4;
+    public readonly EPAISSEUR: number = 40;
+
+    public readonly LEFT: number = 0;
+    public readonly UP: number = -100;
+    public readonly WIDTH: number = (7 * this.SPACE_SIZE) * (this.HORIZONTAL_WIDTH_RATIO + this.DIAGONALISATION);
+    public readonly HEIGHT: number = (7 * this.SPACE_SIZE) + this.EPAISSEUR - this.UP;
+    public readonly CX: number = this.WIDTH / 2;
+    public readonly CY: number = (this.HEIGHT + this.UP) / 2;
+
+    public adaptedBoard: SpaceInfo[][] = [];
+    private lastMove: MGPOptional<LascaMove> = MGPOptional.empty();
+    private validClicks: Coord[] = [];
+    private captureds: Coord[] = [];
+    private validCaptures: LascaMove[] = [];
+    private validSteps: LascaMove[] = [];
+    public LascaSpace: typeof LascaSpace = LascaSpace;
+
+    public constructor(messageDisplayer: MessageDisplayer) {
+        super(messageDisplayer);
+        this.hasAsymetricBoard = true;
+        this.rules = new LascaRules(LascaState);
+        this.availableMinimaxes = [
+            new LascaControlMinimax(this.rules, 'Control Minimax'),
+            new LascaControlAndDominationMinimax(this.rules, 'Control and Domination'),
+        ];
+        this.encoder = LascaMove.encoder;
+        this.tutorial = new LascaTutorial().tutorial;
+        this.canPass = false;
+        this.updateBoard();
+    }
+    public backgroundColor(x: number, y: number): string {
+        if ((x + y) % 2 === 0) {
+            return 'lightgrey';
+        } else {
+            return 'black';
+        }
+    }
+    public updateBoard(): void {
+        this.lastMove = this.rules.node.move;
+        const state: LascaState = this.getState();
+        this.board = state.getCopiedBoard();
+        this.validCaptures = LascaRules.getCaptures(state);
+        this.validSteps = LascaRules.getSteps(state);
```

Comment:
vaudrait mieux passer par le singleton, après tout c'est le but de la singletonisation

Reply:
Vaut mieux pour quoi ?

Reply:
l'avantage du singleton serait d'être sûr d'avoir un seul objet ET de pas devoir passer la classe ou l'instance en paramètre dans le futur, et que tout un chacun sache aller le chercher !

Voilà, adapté.
Note: "this.rules = " apparaît 29 fois dans Lasca, dont 19 suivi de "new" et les 10 autres font Classe.get();, les 19 premiers n'ont pas de méthode get, pas singletonisés du tout.

Reply:
oui mais on pourrait tout faire en static alors ! L'avantage du singleton supplémentaire c'est qu'on peut définir des méthodes normales, qui doivent suivre la légalité dans l'héritage ! 

Et oui y'a encore pas mal de jeux à singletoniser :p
TPR